File: android/app/src/main/java/com/foliole/android/FolioleCompanionAttachmentRetirementJournal.java
package com.foliole.android;

import android.content.Context;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.UUID;

final class FolioleCompanionAttachmentRetirementJournal {
    private FolioleCompanionAttachmentRetirementJournal() {}

    static JSObject prepare(Context context, PluginCall call) throws Exception {
        JSArray tombstones = call.getArray("tombstones");
        if (tombstones == null || tombstones.length() == 0) throw new IllegalArgumentException("tombstones is required");
        String libraryScope = required(call.getData(), "library_scope");
        File root = root(context), stageRoot = new File(root, UUID.randomUUID().toString());
        if (!stageRoot.mkdirs() && !stageRoot.isDirectory()) throw new IllegalStateException("Retirement stage unavailable.");
        JSONArray items = new JSONArray();
        for (int index = 0; index < tombstones.length(); index++) {
            JSONObject value = tombstones.getJSONObject(index);
            String attachmentId = required(value, "attachment_id");
            String contentHash = required(value, "content_hash");
            String storageKey = safeName(required(value, "storage_key"));
            String mimeType = required(value, "mime_type");
            if (!contentHash.matches("^[a-f0-9]{64}$")) throw new IllegalArgumentException("Invalid attachment hash.");
            File source = new File(new File(context.getFilesDir(), "attachments"), storageKey);
            if (source.isFile() && !contentHash.equals(FolioleCompanionAttachmentResourceHash.digestHex(context, source))) {
                throw new IllegalStateException("Attachment retirement source hash mismatch.");
            }
            items.put(new JSONObject().put("attachment_id", attachmentId).put("content_hash", contentHash)
                .put("storage_key", storageKey).put("mime_type", mimeType).put("source", source.getAbsolutePath())
                .put("staged", new File(stageRoot, storageKey).getAbsolutePath()).put("source_present", source.isFile()));
        }
        File journal = new File(root, stageRoot.getName() + ".json");
        JSONObject value = new JSONObject().put("version", 1).put("library_scope", libraryScope)
            .put("stage", "planned").put("stage_history", new JSONArray().put("planned")).put("items", items);
        write(journal, value);
        advance(journal, value, "targets_prepared");
        return new JSObject().put("journal_token", stageRoot.getName());
    }

    static JSObject finish(Context context, PluginCall call) throws Exception {
        String token = safeName(call.getString("journal_token", ""));
        File journal = new File(root(context), token + ".json");
        JSONObject value = new JSONObject(new String(Files.readAllBytes(journal.toPath()), StandardCharsets.UTF_8));
        if (!call.getBoolean("committed", false)) {
            restore(value);
            advance(journal, value, "restored");
            return new JSObject();
        }
        advance(journal, value, "database_committed");
        JSONArray items = value.getJSONArray("items");
        for (int index = 0; index < items.length(); index++) stage(context, items.getJSONObject(index));
        for (int index = 0; index < items.length(); index++) verify(context, items.getJSONObject(index));
        advance(journal, value, "verified");
        return new JSObject();
    }

    static JSObject finalizeJournal(Context context, PluginCall call) throws Exception {
        String token = safeName(call.getString("journal_token", ""));
        File journal = new File(root(context), token + ".json");
        JSONObject value = new JSONObject(new String(Files.readAllBytes(journal.toPath()), StandardCharsets.UTF_8));
        String stage = value.optString("stage");
        if (!stage.equals("verified") && !stage.equals("finalized")) {
            throw new IllegalStateException("Attachment retirement journal is not verified.");
        }
        JSONArray items = value.getJSONArray("items");
        for (int index = 0; index < items.length(); index++) {
            Files.deleteIfExists(new File(items.getJSONObject(index).getString("staged")).toPath());
        }
        advance(journal, value, "finalized");
        return new JSObject();
    }

    private static void stage(Context context, JSONObject item) throws Exception {
        if (!item.getBoolean("source_present")) return;
        File source = new File(item.getString("source")), staged = new File(item.getString("staged"));
        if (staged.isFile()) {
            if (!item.getString("content_hash").equals(FolioleCompanionAttachmentResourceHash.digestHex(context, staged))) {
                throw new IllegalStateException("Attachment retirement staged hash mismatch.");
            }
            return;
        }
        if (!source.isFile() || !item.getString("content_hash").equals(
            FolioleCompanionAttachmentResourceHash.digestHex(context, source))) {
            throw new IllegalStateException("Attachment retirement identity changed.");
        }
        Files.move(source.toPath(), staged.toPath(), StandardCopyOption.ATOMIC_MOVE);
    }

    private static void verify(Context context, JSONObject item) throws Exception {
        if (!item.getBoolean("source_present")) return;
        File source = new File(item.getString("source")), staged = new File(item.getString("staged"));
        if (source.exists() || !staged.isFile() || !item.getString("content_hash").equals(
            FolioleCompanionAttachmentResourceHash.digestHex(context, staged))) {
            throw new IllegalStateException("Attachment retirement staging verification failed.");
        }
    }

    private static void restore(JSONObject value) throws Exception {
        JSONArray items = value.getJSONArray("items");
        for (int index = 0; index < items.length(); index++) {
            JSONObject item = items.getJSONObject(index);
            File source = new File(item.getString("source")), staged = new File(item.getString("staged"));
            if (staged.isFile() && !source.exists()) Files.move(staged.toPath(), source.toPath(), StandardCopyOption.ATOMIC_MOVE);
        }
    }

    private static File root(Context context) {
        File root = new File(context.getFilesDir(), "attachment-retirement");
        if (!root.mkdirs() && !root.isDirectory()) throw new IllegalStateException("Retirement journal unavailable.");
        return root;
    }

    private static String safeName(String value) {
        if (value.isEmpty() || value.contains("/") || value.contains("\\") || value.equals(".") || value.equals("..")) {
            throw new IllegalArgumentException("Invalid retirement path component.");
        }
        return value;
    }

    private static String required(JSONObject value, String key) throws Exception {
        String result = value.optString(key, "").trim();
        if (result.isEmpty()) throw new IllegalArgumentException("Missing " + key);
        return result;
    }

    private static void write(File target, JSONObject value) throws Exception {
        File temporary = new File(target.getParentFile(), target.getName() + ".tmp");
        try (FileOutputStream output = new FileOutputStream(temporary)) {
            output.write((value.toString(2) + "\n").getBytes(StandardCharsets.UTF_8));
            output.getFD().sync();
        }
        Files.move(temporary.toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
    }

    private static void advance(File journal, JSONObject value, String stage) throws Exception {
        JSONArray history = value.optJSONArray("stage_history");
        if (history == null) history = new JSONArray();
        if (history.length() == 0 || !stage.equals(history.optString(history.length() - 1))) history.put(stage);
        write(journal, value.put("stage", stage).put("stage_history", history));
    }
}

File: ios/App/App/FolioleCompanionAttachmentRetirementJournal.swift
import Capacitor
import Foundation

enum FolioleCompanionAttachmentRetirementJournal {
    static func prepare(_ call: CAPPluginCall, directoryName: String) throws -> [String: Any] {
        guard let raw = call.getArray("tombstones"), !raw.isEmpty else { throw invalid("tombstones is required") }
        guard let libraryScope = call.getString("library_scope"), !libraryScope.isEmpty else {
            throw invalid("library_scope is required")
        }
        let root = try journalRoot(), token = UUID().uuidString.lowercased()
        let stageRoot = root.appendingPathComponent(token, isDirectory: true)
        try FileManager.default.createDirectory(at: stageRoot, withIntermediateDirectories: true)
        let attachmentRoot = try supportRoot().appendingPathComponent(directoryName, isDirectory: true)
        let items = try raw.map { value -> [String: Any] in
            guard let item = value as? [String: Any] else { throw invalid("Invalid attachment tombstone") }
            let attachmentId = try required(item, "attachment_id")
            let contentHash = try required(item, "content_hash")
            let storageKey = try safeName(required(item, "storage_key"))
            let mimeType = try required(item, "mime_type")
            guard contentHash.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil else {
                throw invalid("Invalid attachment hash")
            }
            let source = attachmentRoot.appendingPathComponent(storageKey)
            let present = FileManager.default.fileExists(atPath: source.path)
            if present {
                let actualHash = try FolioleCompanionAttachmentResourceDownloader.digestHex(source)
                if actualHash != contentHash { throw invalid("Attachment retirement source hash mismatch") }
            }
            return ["attachment_id": attachmentId, "content_hash": contentHash, "storage_key": storageKey,
                    "mime_type": mimeType, "source": source.path,
                    "staged": stageRoot.appendingPathComponent(storageKey).path, "source_present": present]
        }
        let journal = root.appendingPathComponent("\(token).json")
        var value: [String: Any] = ["version": 1, "library_scope": libraryScope,
                                    "stage": "planned", "stage_history": ["planned"], "items": items]
        try write(value, journal)
        try advance(&value, journal, "targets_prepared")
        return ["journal_token": token]
    }

    static func finish(_ call: CAPPluginCall) throws {
        let token = try safeName(call.getString("journal_token") ?? "")
        let journalURL = try journalRoot().appendingPathComponent("\(token).json")
        var journal = try read(journalURL)
        guard let items = journal["items"] as? [[String: Any]] else { throw invalid("Invalid retirement journal") }
        if call.getBool("committed") != true {
            try restore(items)
            try advance(&journal, journalURL, "restored")
            return
        }
        try advance(&journal, journalURL, "database_committed")
        for item in items { try stage(item) }
        for item in items { try verify(item) }
        try advance(&journal, journalURL, "verified")
    }

    static func finalize(_ call: CAPPluginCall) throws {
        let token = try safeName(call.getString("journal_token") ?? "")
        let journalURL = try journalRoot().appendingPathComponent("\(token).json")
        var journal = try read(journalURL)
        let stage = journal["stage"] as? String
        guard stage == "verified" || stage == "finalized",
              let items = journal["items"] as? [[String: Any]] else { throw invalid("Journal is not verified") }
        for item in items {
            guard let stagedPath = item["staged"] as? String else { throw invalid("Invalid staged path") }
            let staged = URL(fileURLWithPath: stagedPath)
            if FileManager.default.fileExists(atPath: staged.path) { try FileManager.default.removeItem(at: staged) }
        }
        try advance(&journal, journalURL, "finalized")
    }

    private static func stage(_ item: [String: Any]) throws {
        guard item["source_present"] as? Bool == true,
              let sourcePath = item["source"] as? String,
              let stagedPath = item["staged"] as? String,
              let contentHash = item["content_hash"] as? String else { return }
        let source = URL(fileURLWithPath: sourcePath), staged = URL(fileURLWithPath: stagedPath)
        if FileManager.default.fileExists(atPath: staged.path) {
            let stagedHash = try FolioleCompanionAttachmentResourceDownloader.digestHex(staged)
            guard stagedHash == contentHash else {
                throw invalid("Attachment retirement staged hash mismatch")
            }
            return
        }
        guard FileManager.default.fileExists(atPath: source.path),
              try FolioleCompanionAttachmentResourceDownloader.digestHex(source) == contentHash else {
            throw invalid("Attachment retirement identity changed")
        }
        try FileManager.default.moveItem(at: source, to: staged)
    }

    private static func verify(_ item: [String: Any]) throws {
        guard item["source_present"] as? Bool == true,
              let sourcePath = item["source"] as? String,
              let stagedPath = item["staged"] as? String,
              let contentHash = item["content_hash"] as? String else { return }
        let source = URL(fileURLWithPath: sourcePath), staged = URL(fileURLWithPath: stagedPath)
        let stagedHash = try FolioleCompanionAttachmentResourceDownloader.digestHex(staged)
        guard !FileManager.default.fileExists(atPath: source.path),
              FileManager.default.fileExists(atPath: staged.path),
              stagedHash == contentHash else {
            throw invalid("Attachment retirement staging verification failed")
        }
    }

    private static func restore(_ items: [[String: Any]]) throws {
        for item in items {
            guard let sourcePath = item["source"] as? String,
                  let stagedPath = item["staged"] as? String else { continue }
            let source = URL(fileURLWithPath: sourcePath), staged = URL(fileURLWithPath: stagedPath)
            if FileManager.default.fileExists(atPath: staged.path),
               !FileManager.default.fileExists(atPath: source.path) {
                try FileManager.default.moveItem(at: staged, to: source)
            }
        }
    }

    private static func supportRoot() throws -> URL {
        try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask,
                                    appropriateFor: nil, create: true)
    }

    private static func journalRoot() throws -> URL {
        let root = try supportRoot().appendingPathComponent("attachment-retirement", isDirectory: true)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        return root
    }

    private static func read(_ url: URL) throws -> [String: Any] {
        guard let value = try JSONSerialization.jsonObject(with: Data(contentsOf: url)) as? [String: Any] else {
            throw invalid("Invalid retirement journal")
        }
        return value
    }

    private static func write(_ value: [String: Any], _ url: URL) throws {
        let data = try JSONSerialization.data(withJSONObject: value, options: [.prettyPrinted, .sortedKeys])
        try data.write(to: url, options: .atomic)
    }

    private static func advance(_ value: inout [String: Any], _ url: URL, _ stage: String) throws {
        var history = value["stage_history"] as? [String] ?? []
        if history.last != stage { history.append(stage) }
        value["stage"] = stage
        value["stage_history"] = history
        try write(value, url)
    }

    private static func required(_ value: [String: Any], _ key: String) throws -> String {
        guard let result = value[key] as? String, !result.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw invalid("Missing \(key)")
        }
        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func safeName(_ value: String) throws -> String {
        guard !value.isEmpty, value != ".", value != "..", !value.contains("/"), !value.contains("\\") else {
            throw invalid("Invalid retirement path component")
        }
        return value
    }

    private static func invalid(_ message: String) -> NSError {
        NSError(domain: "FolioleCompanionAttachmentRetirementJournal", code: 1,
                userInfo: [NSLocalizedDescriptionKey: message])
    }
}
